hazelcast/src/main/java/com/hazelcast/sql/impl/worker/QueryFragmentWorkerPool.java

```diff
@@ -0,0 +1,80 @@
+/*
+ * Copyright (c) 2008-2020, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.sql.impl.worker;
+
+import com.hazelcast.sql.impl.QueryUtils;
+
+import java.util.concurrent.ForkJoinPool;
+import java.util.concurrent.ForkJoinWorkerThread;
+import java.util.concurrent.atomic.AtomicLong;
+
+import static com.hazelcast.sql.impl.QueryUtils.WORKER_TYPE_FRAGMENT;
+
+/**
+ * Thread pool that executes query fragments.
+ */
+public class QueryFragmentWorkerPool {
+
+    private final ForkJoinPool pool;
+
+    public QueryFragmentWorkerPool(String instanceName, int threadCount) {
+        pool = new ForkJoinPool(threadCount, new WorkerThreadFactory(instanceName), null, false);
```

Comment:
It's not a good idea to pass `null` as `UncaughtExceptionHandler`, all exceptions will be silenced.

Reply:
... `QueryOperationWorkerPool` also silences all exceptions.

Reply:
We do not need anything to handle here since exceptions are handled in other places (e.g. `QueryFragmentExecutable.run()`). 
Even JDK doesn't use any handler by default.

Reply:
JDK doesn't handle uncaught exceptions because it doesn't know what to do with them, there is no good default strategy that works for everyone. To decide on the strategy they provide an extension point.

`QueryFragmentExecutable` doesn't handle all exceptions, even basic failed `assert` is swallowed silently because it results in `AssertionError`, which is not a subclass of `Exception`. Besides that, the `unschedule()` call is not wrapped in try-catch. In any case, it's a very bad practice to leave uncaught exception handlers empty.

Reply:
Note these are not generic thread pools that execute arbitrary tasks. The semantics is fully under our control. Moreover, FJP re-creates failed threads, when something nasty happens. What kind of handling do you suggest to add here?

I think that one thing that is definitely worth adding is "panic" handler, such as in `com.hazelcast.spi.impl.operationexecutor.impl.OperationThread#run`, but it should be installed on the thread level, not thread pool.

Reply:
Yes, it's under our control, but that doesn't mean we should treat all the code base as fully verified and 100% bug free. We can even invoke user code under the hood: every map scan potentially results in accessing user supplied classes, we read attribute values from them.

The usual practice is to at least log the errors you can't handle, looks like we can't do nothing more.

Reply:
The `asyncMode` parameter is basically a switch between FIFO and LIFO modes, `false` = LIFO. It should not affect the execution, since we don't fork any tasks, but `Executors.newWorkStealingPool` sets the parameter to `true`, probably we should do the same.

Reply:
Added exception handling to both pools. It follows the common approach we use in other pools - to log an error, and possibly handle OOME. It is far from ideal, but it is what it is.

Reply:
Regarding `asyncMode` - great catch, LIFO is dangerous for use because it may lead to starvation. Changed to `true`.